File: src/main/java/com/plotsx/models/Plot.java
package com.plotsx.models;

import org.bukkit.Location;
import java.util.HashSet;
import java.util.Set;
import java.util.UUID;

public class Plot {
    private final UUID id;
    private final UUID owner;
    private final Set<UUID> coOwners;
    private final Location center;
    private final int radius;
    private final long creationDate;
    private boolean active;
    private boolean pvpEnabled;
    private boolean mobDamageEnabled;
    private boolean basicInteractionsEnabled;

    public Plot(UUID owner, Location center, int radius) {
        this.id = UUID.randomUUID();
        this.owner = owner;
        this.coOwners = new HashSet<>();
        this.center = center;
        this.radius = radius;
        this.creationDate = System.currentTimeMillis();
        this.active = true;
        this.pvpEnabled = false;
        this.mobDamageEnabled = false;
        this.basicInteractionsEnabled = true;
    }

    public UUID getId() {
        return id;
    }

    public UUID getOwner() {
        return owner;
    }

    public Set<UUID> getCoOwners() {
        return coOwners;
    }

    public Location getCenter() {
        return center;
    }

    public int getRadius() {
        return radius;
    }

    public long getCreationDate() {
        return creationDate;
    }

    public boolean isActive() {
        return active;
    }

    public void setActive(boolean active) {
        this.active = active;
    }

    public boolean isPvpEnabled() {
        return pvpEnabled;
    }

    public void setPvpEnabled(boolean pvpEnabled) {
        this.pvpEnabled = pvpEnabled;
    }

    public boolean isMobDamageEnabled() {
        return mobDamageEnabled;
    }

    public void setMobDamageEnabled(boolean mobDamageEnabled) {
        this.mobDamageEnabled = mobDamageEnabled;
    }

    public boolean isBasicInteractionsEnabled() {
        return basicInteractionsEnabled;
    }

    public void setBasicInteractionsEnabled(boolean basicInteractionsEnabled) {
        this.basicInteractionsEnabled = basicInteractionsEnabled;
    }

    public boolean isOwner(UUID playerId) {
        return owner.equals(playerId);
    }

    public boolean isCoOwner(UUID playerId) {
        return coOwners.contains(playerId);
    }

    public boolean hasAccess(UUID playerId) {
        return isOwner(playerId) || isCoOwner(playerId);
    }

    public void addCoOwner(UUID playerId) {
        coOwners.add(playerId);
    }

    public void removeCoOwner(UUID playerId) {
        coOwners.remove(playerId);
    }

    public boolean isInPlot(Location location) {
        if (!location.getWorld().equals(center.getWorld())) {
            return false;
        }
        
        double distanceX = Math.abs(location.getX() - center.getX());
        double distanceZ = Math.abs(location.getZ() - center.getZ());
        
        return distanceX <= radius && distanceZ <= radius;
    }
} 

File: src/main/kotlin/com/plotsx/models/Plot.kt
package com.plotsx.models

import org.bukkit.Location
import org.bukkit.entity.Player
import java.util.*
import kotlin.math.abs

data class Plot(
    val id: String,
    val owner: UUID,
    val world: String,
    val x1: Int,
    val y1: Int,
    val z1: Int,
    val x2: Int,
    val y2: Int,
    val z2: Int,
    val coowners: Set<UUID> = emptySet(),
    val flags: Map<String, Any> = emptyMap()
) {
    val center: Location by lazy {
        Location(
            org.bukkit.Bukkit.getWorld(world),
            (x1 + x2) / 2.0,
            (y1 + y2) / 2.0,
            (z1 + z2) / 2.0
        )
    }

    val corners: List<Location> by lazy {
        listOf(
            Location(org.bukkit.Bukkit.getWorld(world), x1.toDouble(), y1.toDouble(), z1.toDouble()),
            Location(org.bukkit.Bukkit.getWorld(world), x2.toDouble(), y1.toDouble(), z1.toDouble()),
            Location(org.bukkit.Bukkit.getWorld(world), x1.toDouble(), y1.toDouble(), z2.toDouble()),
            Location(org.bukkit.Bukkit.getWorld(world), x2.toDouble(), y1.toDouble(), z2.toDouble())
        )
    }

    operator fun contains(location: Location): Boolean = isInPlot(location)
    
    operator fun contains(player: Player): Boolean = isInPlot(player.location)

    fun isOwner(playerId: UUID): Boolean = owner == playerId

    fun isCoOwner(playerId: UUID): Boolean = coowners.contains(playerId)

    fun hasAccess(playerId: UUID): Boolean = isOwner(playerId) || isCoOwner(playerId)

    fun isInPlot(location: Location): Boolean {
        if (location.world?.name != world) return false
        
        return location.blockX in x1..x2 && 
               location.blockY in y1..y2 && 
               location.blockZ in z1..z2
    }

    fun getFlag(key: String): Any? = flags[key]

    fun hasFlag(key: String): Boolean = flags.containsKey(key)

    companion object {
        fun fromConfig(config: Map<String, Any>): Plot {
            return Plot(
                id = config["id"] as String,
                owner = UUID.fromString(config["owner"] as String),
                world = config["world"] as String,
                x1 = config["x1"] as Int,
                y1 = config["y1"] as Int,
                z1 = config["z1"] as Int,
                x2 = config["x2"] as Int,
                y2 = config["y2"] as Int,
                z2 = config["z2"] as Int,
                coowners = (config["coowners"] as List<String>).map { UUID.fromString(it) }.toSet(),
                flags = config["flags"] as? Map<String, Any> ?: emptyMap()
            )
        }
    }
}

// Extension functions
fun Plot.sendMessageToOwners(message: String) {
    org.bukkit.Bukkit.getWorld(world)?.let { world ->
        world.onlinePlayers
            .filter { player -> hasAccess(player.uniqueId) }
            .forEach { player -> player.sendMessage(message) }
    }
}

fun Plot.getOnlineOwners(): List<Player> {
    return org.bukkit.Bukkit.getWorld(world)?.onlinePlayers
        ?.filter { player -> hasAccess(player.uniqueId) }
        ?: emptyList()
}

fun Plot.toConfig(): Map<String, Any> {
    return mapOf(
        "id" to id,
        "owner" to owner.toString(),
        "world" to world,
        "x1" to x1,
        "y1" to y1,
        "z1" to z1,
        "x2" to x2,
        "y2" to y2,
        "z2" to z2,
        "coowners" to coowners.map { it.toString() },
        "flags" to flags
    )
} 